cache/volume_cache.go

```
@@ -0,0 +1,145 @@
+		return errors.Wrapf(err, "creating metadata file '%s'", metadataPath)
+	}
+	defer file.Close()
+
+	if err := json.NewEncoder(file).Encode(metadata); err != nil {
+		return errors.Wrap(err, "marshalling metadata")
+	}
+
+	return nil
+}
+
+func (c *VolumeCache) RetrieveMetadata() (Metadata, error) {
+	metadataPath := filepath.Join(c.committedDir, MetadataLabel)
+	file, err := os.Open(metadataPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return Metadata{}, nil
+		}
+		return Metadata{}, errors.Wrapf(err, "opening metadata file '%s'", metadataPath)
+	}
+	defer file.Close()
+
+	metadata := Metadata{}
+	_ = json.NewDecoder(file).Decode(&metadata)
+	return metadata, nil
+}
+
+func (c *VolumeCache) AddLayer(identifier string, sha string, tarPath string) error {
+	if err := copyFile(tarPath, filepath.Join(c.stagingDir, sha+".tar")); err != nil {
+		return errors.Wrapf(err, "caching layer '%s' (%s)", identifier, sha)
+	}
+	return nil
+}
+
+func (c *VolumeCache) ReuseLayer(identifier string, sha string) error {
+	if err := copyFile(filepath.Join(c.committedDir, sha+".tar"), filepath.Join(c.stagingDir, sha+".tar")); err != nil {
+		return errors.Wrapf(err, "reusing layer '%s' (%s)", identifier, sha)
+	}
+	return nil
+}
+
+func (c *VolumeCache) RetrieveLayer(sha string) (io.ReadCloser, error) {
+	file, err := os.Open(filepath.Join(c.committedDir, sha+".tar"))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, errors.Wrapf(err, "layer with SHA '%s' not found", sha)
+		}
+		return nil, errors.Wrapf(err, "retrieving layer with SHA '%s'", sha)
+	}
+	return file, nil
+}
+
+func (c *VolumeCache) Commit() error {
+	if err := os.Rename(c.committedDir, c.backupDir); err != nil {
+		return errors.Wrap(err, "backing up cache")
+	}
+	defer os.RemoveAll(c.backupDir)
+
+	if err := os.Rename(c.stagingDir, c.committedDir); err != nil {
+		if err := os.Rename(c.backupDir, c.committedDir); err != nil {
+			return errors.Wrap(err, "rolling back cache")
+		}
+		return nil
```

Are we ignoring the error intentionally when the rename fails? If so, what condition are we trying to ignore?

---

Yes, if committing the cache fails, we attempt a rollback (no error). If the rollback fails, we error.

---

👍